Add blog post to announce ApplySet-based pruning in kubectl (#40070)
kubernetes/website: content/en/blog/_posts/2023-04-11-Safer-More-Performant-Pruning-in-Kubectl-Apply.md

---
layout: blog

> title: "Safer, More Performant Pruning in Kubectl Apply"

Reviewer: ```suggestion
title: "Kubernetes 1.27: Safer, More Performant Pruning in kubectl apply"
```

date: 2023-04-11
slug: introducing-kubectl-applyset-pruning
---

**Authors:** Katrina Verey (Shopify) and Justin Santa Barbara (Google)

Declarative configuration management with the `kubectl apply` command is the gold standard approach
to creating or modifying Kubernetes resources. However, one challenge it presents is the deletion
of resources that are no longer needed. In Kubernetes version 1.5, the `--prune` flag was
introduced to address this issue, allowing kubectl apply to automatically clean up previously
applied resources removed from the current configuration.

> Unfortunately, the current implementation of `--prune` has design flaws that diminish its

Reviewer: ```suggestion
Unfortunately, that existing implementation of `--prune` has design flaws that diminish its
```

its coupling to client-side apply hinders user upgrades to the superior server-side apply
mechanism.

Version 1.27 of `kubectl` introduces an alpha version of a revamped pruning implementation that
addresses these issues. This new implementation, based on a concept called _ApplySet_, promises
better performance and safety.

An _ApplySet_ is a group of resources associated with a _parent_ object on the cluster, as
identified and configured through standardized labels and annotations. Additional standardized
metadata allows for accurate identification of ApplySet _member_ objects within the cluster,
simplifying operations like pruning.

To leverage ApplySet-based pruning, set the `KUBECTL_APPLYSET=true` environment variable and include
the flags `--prune` and `--applyset` in your `kubectl apply` invocation:

```shell
KUBECTL_APPLYSET=true kubectl apply -f <directory/> --prune --applyset=<name>
```

By default, ApplySet uses a `Secret` resource as the parent object. However, you can also use
`ConfigMap` resource with the format `--applyset=configmaps/<name>`. If your desired `Secret` or
`ConfigMap` object does not yet exist, `kubectl` will create it for you. Furthermore, custom

> resources can be enabled for use as ApplySet parent objects.

Author: I have a question about this one--all of these changes were actually things I did after reading the style guide, which I read as saying the the word "resource" or "object" should always be used with types, and that types should be styled as code. Here are the two sections in question:
- https://kubernetes.io/docs/contribute/style/style-guide/#code-style-inline-code
- https://kubernetes.io/docs/contribute/style/style-guide/#when-to-use-kubernetes-api-terminologies

> resources can be enabled for use as ApplySet parent objects.

Reviewer: The style guide is not strict for blog articles, and we need to fix our style guide advice about using monospace for API kinds.

The problem is plurals, eg:  “the controller watches for `PersistentVolumeClaim`s and then provisions `PersistentVolume`s to match the `StorageClass` as well as the `CSIDriver`'s requirements (if relevant)”. Which kinda looks weird.

> resources can be enabled for use as ApplySet parent objects.

Author: Ok, updated.

The ApplySet implementation is based on a new low-level specification that can support higher-level
ecosystem tools by improving their interoperability. The lightweight nature of this specification
enables these tools to continue to use existing object grouping systems while opting in to
ApplySet's metadata conventions to prevent inadvertent changes by other tools (such as `kubectl`).

ApplySet-based pruning offers a promising solution to the shortcomings of the previous `--prune`
implementation in `kubectl` and can help streamline your Kubernetes resource management. Please

> give this new feature a try and share your experiences with the community--ApplySet is under active

Reviewer: ```suggestion
give this new feature a try and share your experiences with the community—ApplySet is under active
```
